scheduler/src/cook/mesos/util.clj

```diff
@@ -198,38 +199,73 @@
+               (seq batches))
+        (recur coll' batches)
+        (take n coll')))))
+
 (defn get-jobs-by-user-and-state
   "Returns all job entities for a particular user
    in a particular state, in the specified timeframe,
    without a custom executor."
-  [db user state start end]
-  (->> (if (= state :job.state/completed)
-         ;; Datomic query performance is based entirely on the size of the set
-         ;; of the first where clause. In the case of :job.state/completed
-         ;; the total number of jobs for a user should be smaller than
-         ;; all completed jobs by any user i.e.
-         ;; (waiting_user + running_user < completed - completed_user)
-         (q '[:find [?j ...]
-              :in $ ?user ?state ?start ?end
-              :where
-              [?j :job/user ?user]
-              [?j :job/state ?state]
-              [?j :job/submit-time ?t]
-              [(< ?start ?t)]
-              [(< ?t ?end)]
-              [?j :job/custom-executor false]]
-            db user state start end)
-         (q '[:find [?j ...]
-              :in $ ?user ?state ?start ?end
-              :where
-              [?j :job/state ?state]
-              [?j :job/user ?user]
-              [?j :job/submit-time ?t]
-              [(< ?start ?t)]
-              [(< ?t ?end)]
-              [?j :job/custom-executor false]]
-            db user state start end))
-       (map (partial d/entity db))))
+  [db user state start end limit]
+  (let [intervals (generate-intervals (tc/from-date start) (tc/from-date end) (t/days 1))
+        job-batches (if (= state :job.state/completed)
+                      (map (fn query-in-chunks
+                             [interval]
+                             (let [[interval-start interval-end] (map tc/to-date interval)]
+                               (q '[:find [?j ...]
+                                    :in $ ?user ?state ?start ?end
+                                    :where
+                                    [?j :job/submit-time ?t]
+                                    [(<= ?start ?t)]
+                                    [(< ?t ?end)]
+                                    [?j :job/user ?user]
+                                    [?j :job/state ?state]
+                                    [?j :job/custom-executor false]]
+                                  db user state interval-start interval-end)))
+                           intervals)
+                      [(q '[:find [?j ...]
+                             :in $ ?user ?state ?start ?end
+                             :where
+                             [?j :job/state ?state]
+                             [?j :job/user ?user]
+                             [?j :job/submit-time ?t]
+                             [(<= ?start ?t)]
+                             [(< ?t ?end)]
+                             [?j :job/custom-executor false]]
+                           db user state start end)])]
+    (->> job-batches
+         (map sort)
```

Comment:
Why do the batches need to be sorted? Is this intended to be applied on the accumulated results instead?